### PR TITLE
Added FSRS info to getDebugInfo

### LINKED ISSUE #15447: Add scheduler/version status to debug info
We want to know if the user is using FSRS when we get a debug report

https://github.com/ankidroid/Anki-Android/blob/5c75c820d6c2f262f104aad6ac088fc982ba9897/AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt#L29-L52

### PR DESCRIPTION
**Pull request by Haz3-jolt**
## Purpose / Description
Added FSRS status to getDebugInfo as a coroutine and converted getDebugInfo into a suspend function and converted it's callers to applicationScopeLaunch.

## Fixes
* Fixes #15447

## Approach
- Adds info to the debug about weather the user is using FSRS or not, by creating a suspend fun getFSRSStatus(), which uses config.get() to get the FSRS status. 
- Converted getDebugInfo() to a suspend function.
- Wrapped all the relevant getDebugInfo() with a coroutine. 
## How Has This Been Tested?

When through compilation, tested by using a fresh install and running cases the following cases: 
- With no decks (Fresh Install). (FSRS = null)
- One deck with no cards. (FSRS = null)
- One deck with Cards but FSRS has never been enabled. (FSRS = null)
- One deck with Cards and FSRS has been enabled. (FSRS = True)
- One deck with Card and FSRS has been disabled but was enabled some time in the past. (FSRS = False)
- With no decks but FSRS has been enabled before. (FSRS = null)
- New empty deck created with FSRS never enabled. (FSRS = False)

## Checklist
- [-] You have a descriptive commit message with a short title (first line, max 50 chars).
- [-] You have commented your code, particularly in hard-to-understand areas
- [-] You have performed a self-review of your own code
- [-] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [-] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

Here is a video with some TestCases
https://github.com/ankidroid/Anki-Android/assets/79502699/8184ce57-9c9a-4580-bd45-510690537073

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.kt [hunk 1]
```diff
@@ -133,8 +134,15 @@ open class AnkiDroidApp : Application() {
         if (BuildConfig.DEBUG) {
             UsageAnalytics.setDryRun(true)
         }
-
-        Timber.i(DebugInfoService.getDebugInfo(this))
+        // Use applicationScope as a coroutine scope to use getDebugInfo() since it is a suspend fun
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.kt:137]: the comment isn't useful

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.kt [hunk 2]
```diff
@@ -133,8 +134,15 @@ open class AnkiDroidApp : Application() {
...
+        applicationScope.launch {
+            try {
+                val debugInfo = DebugInfoService.getDebugInfo(this@AnkiDroidApp)
+                Timber.i(debugInfo)
+            } catch (e: Exception) {
+                Timber.e(e, "Error getting debug info")
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.kt:143]:
```suggestion
                Timber.w(e, "Error getting debug info")
```

File: AnkiDroid/src/main/java/com/ichi2/anki/preferences/AboutFragment.kt [hunk 3]
```diff
@@ -111,11 +116,20 @@ class AboutFragment : Fragment(R.layout.about_layout) {
      * Copies debug info (from [DebugInfoService.getDebugInfo]) to the clipboard
      */
     private fun copyDebugInfo() {
-        val debugInfo = DebugInfoService.getDebugInfo(requireContext())
-        requireContext().copyToClipboard(
-            debugInfo,
-            failureMessageId = R.string.about_ankidroid_error_copy_debug_info
-        )
+        // Use lifecycleScope as a coroutine scope to use getDebugInfo() since it is a suspend fun
+        viewLifecycleOwner.lifecycleScope.launch {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/preferences/AboutFragment.kt:120]: Use `launchCatchingTask` and drop the `catch` block

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt [hunk 4]
```diff
@@ -20,14 +20,28 @@ import android.content.Context
...
+
+    // This is run on the startup path, we need to protect against a null/corrupt collection
+    suspend fun getDebugInfo(info: Context): String {
         val webviewUserAgent = getWebviewUserAgent(info)
+        // Launch a coroutine to get the FSRS status
+        val deferred: Deferred<Boolean?> = CoroutineScope(Dispatchers.Main).async {
+            getFSRSStatus()
+        }
+        // Await the result of the coroutine
+        val isFSRSEnabled = deferred.await()
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:44]: Why async/await and why Dispatchers.Main?

**Review by BrayanDSO — CHANGES_REQUESTED** [review]

**Comment by BrayanDSO** [comment]
Also, the commit messages are identical. If they are meant to do the same thing, please squash them

**Comment by Haz3-jolt** [comment]
> Also, the commit messages are identical. If they are meant to do the same thing, please squash them

Ye that was a mistake, will change now.

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt (hunk 4 above)
Comment (Haz3-jolt) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:44]: Was a test case thing, sorry bruv.

File: AnkiDroid/src/main/java/com/ichi2/anki/preferences/AboutFragment.kt [hunk 5]
```diff
@@ -111,11 +114,19 @@ class AboutFragment : Fragment(R.layout.about_layout) {
...
+                val debugInfo = withContext(Dispatchers.IO) {
+                    DebugInfoService.getDebugInfo(requireContext())
+                }
+                requireContext().copyToClipboard(
+                    debugInfo,
+                    failureMessageId = R.string.about_ankidroid_error_copy_debug_info
+                )
+            } catch (e: Exception) {
+                Timber.w(e, "Error copying debug info")
+            }
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/preferences/AboutFragment.kt:128]: not necessary inside a `launchCatchingTask

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt [hunk 6]
```diff
@@ -63,4 +70,12 @@ object DebugInfoService {
     private fun isSendingCrashReports(context: Context): Boolean {
         return CrashReportService.isAcraEnabled(context, false)
     }
+
+    // Uses config.get() to get the FSRS status from the collection using withOpenColOrNull
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:74]: `//` comments can only be read in the file of the method itself. Since this is just stating what the method calls, it is easier to read the method directly.

If it was a `/**` comment, it still wouldn't be necessary to state those kind of implementation details.

TLDR: please remove the comment

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt [hunk 7]
```diff
@@ -47,6 +52,8 @@ object DebugInfoService {
                
                ACRA UUID = ${Installation.id(info)}
                
+               FSRS  = $isFSRSEnabled
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:55]:
```suggestion
               FSRS Enabled = $isFSRSEnabled
```

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt [hunk 8]
```diff
@@ -20,14 +20,19 @@ import android.content.Context
...
 import com.ichi2.anki.CrashReportService
 import com.ichi2.utils.VersionUtils.pkgVersionName
 import org.acra.util.Installation
+import timber.log.Timber
 import net.ankiweb.rsdroid.BuildConfig as BackendBuildConfig
 
 object DebugInfoService {
-    fun getDebugInfo(info: Context): String {
+
+    // This is run on the startup path, we need to protect against a null/corrupt collection
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:32]: A method shouldn't be responsible for its calling sites. The call could be removed from the startup at any time and this wouldn't make sense. Also why someone needs to protect against a null/corrupt collection here?

What I suggest is
```kotlin
    /**
    * Retrieves the debug info based in different parameters of the app.
    * 
    * Note that the `FSRS` parameter can be null if the collection isn't open
    */
```

**Review by BrayanDSO — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt (hunk 8 above)
Comment (Haz3-jolt) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt]: Done, made changes, and the null/corrupt thing was from testing because before I ran the FSRS in a async coroutine since i wanted the go ahead from David on if i can make the entire debugInfo fun into a suspend.

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt (hunk 6 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:78]: > TLDR: please remove the comment

you resolved the comment, but didn't do what was asked nor stated why you didn't

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.kt [hunk 9]
```diff
@@ -134,7 +135,14 @@ open class AnkiDroidApp : Application() {
...
         }
 
-        Timber.i(DebugInfoService.getDebugInfo(this))
+        applicationScope.launch {
+            try {
+                val debugInfo = DebugInfoService.getDebugInfo(this@AnkiDroidApp)
+                Timber.i(debugInfo)
+            } catch (e: Exception) {
+                Timber.w(e, "Error getting debug info")
+            }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.kt:144]: getDebugInfo should not throw

```suggestion
            Timber.i(DebugInfoService.getDebugInfo(this))
```

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt [hunk 10]
```diff
@@ -63,4 +74,12 @@ object DebugInfoService {
     private fun isSendingCrashReports(context: Context): Boolean {
         return CrashReportService.isAcraEnabled(context, false)
     }
+
+    // Uses config.get() to get the FSRS status from the collection using withOpenColOrNull
+    private suspend fun getFSRSStatus(): Boolean? = try {
+        CollectionManager.withOpenColOrNull { config.get<Boolean>("fsrs") }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:80]: If this does not exist, it throws a `BackendNotFoundException`, and therefore returns `null`.

This should be changed s.t.

* `BackendNotFoundException` is caught
* the default value of `false` is returned if the value is not found

**Review by david-allison — CHANGES_REQUESTED** [review]
Thanks!! almost there

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt (hunk 10 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:80]: This could also be implemented as an extension method on `config`, up to you

Comment (Haz3-jolt) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:80]: It returns Null, if there are no decks or if a deck has been created and FSRS has never been enabled, but if FSRS is enabled and then disabled, it would return False. I left it because it might be useful to know if FSRS was used at some point but later disabled, but if you think its better to just have true or false then I am okay with that as well.

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:80]: I feel the most intuitive is:

`null` -> no collection
`false` -> enabled, then disabled (or no preference)
`true` -> enabled

Comment (Haz3-jolt) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:80]: so do i go ahead with the change? btw ignore the checks for a moment, accidentally hit commit suggested changed, have to reverse it rn.

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:80]:
```patch
Subject: [PATCH] feat: copy card template to clipboard

---
Index: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.kt b/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.kt	(revision 7a052a9e0c73dc8f0fd8dc8e7e24af637cf52c19)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.kt	(date 1709419061597)
@@ -136,7 +136,7 @@
         }
 
         applicationScope.launch {
-            Timber.i(DebugInfoService.getDebugInfo(this))
+            Timber.i(DebugInfoService.getDebugInfo(this@AnkiDroidApp))
         }
 
         // Stop after analytics and logging are initialised.
Index: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt b/AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt	(revision 7a052a9e0c73dc8f0fd8dc8e7e24af637cf52c19)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt	(date 1709419363652)
@@ -36,6 +36,7 @@
      */
     suspend fun getDebugInfo(info: Context): String {
         val webviewUserAgent = getWebviewUserAgent(info)
+        // isFSRSEnabled is null on startup
         val isFSRSEnabled = getFSRSStatus()
         return """
                AnkiDroid Version = $pkgVersionName (${BuildConfig.GIT_COMMIT_HASH})
@@ -77,7 +78,7 @@
 
     // Uses config.get() to get the FSRS status from the collection using withOpenColOrNull
     private suspend fun getFSRSStatus(): Boolean? = try {
-        CollectionManager.withOpenColOrNull { config.get<Boolean>("fsrs") }
+        CollectionManager.withOpenColOrNull { config.get<Boolean>("fsrs", false) }
     } catch (e: Error) {
         Timber.w(e)
         null
Index: AnkiDroid/src/main/java/com/ichi2/libanki/Config.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/libanki/Config.kt b/AnkiDroid/src/main/java/com/ichi2/libanki/Config.kt
--- a/AnkiDroid/src/main/java/com/ichi2/libanki/Config.kt	(revision 7a052a9e0c73dc8f0fd8dc8e7e24af637cf52c19)
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/Config.kt	(date 1709419016527)
@@ -18,6 +18,7 @@
 
 import anki.config.ConfigKey
 import com.google.protobuf.kotlin.toByteStringUtf8
+import com.ichi2.libanki.utils.NotInLibAnki
 import kotlinx.serialization.SerializationException
 import kotlinx.serialization.encodeToString
 import kotlinx.serialization.json.Json
@@ -57,4 +58,15 @@
     fun setBool(key: ConfigKey.Bool, value: Boolean) {
         backend.setConfigBool(key, value, false)
     }
+
+    @NotInLibAnki
+    inline fun<reified T> get(key: String, default: T): T? {
+        return try {
+            Json.decodeFromString<T>(backend.getConfigJson(key).toStringUtf8())
+        } catch (ex: BackendNotFoundException) {
+            default
+        } catch (ex: SerializationException) {
+            null
+        }
+    }
 }
```

Comment (Haz3-jolt) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt]: Done.

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt (hunk 6 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:79]: This

**Review by david-allison — CHANGES_REQUESTED** [review]
one comment-based change, then good to go

**Comment by Haz3-jolt** [comment]
> one comment-based change, then good to go

what should the new comment be?

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt (hunk 6 above)
Comment (Haz3-jolt) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/DebugInfoService.kt:79]: ahh ok man

**Review by david-allison — APPROVED** [review]
Looks great, thanks so much @Haz3-jolt!!

**Review by BrayanDSO — APPROVED** [review]
Nice. Thanks!
